docs: Updating Spec (#24)
trustbloc/trustbloc-did-method: docs/overview/trustbloc-did-method.md

# TrustBloc DID Method Specification 0.1

#### Spec Version
This is version `0.1` of the Trustbloc DID Method Specification.

## `did:trustbloc` Method Specification
The `did:trustbloc` DID method allows groups of independent organizations to share custody of a DID registry on a permissioned ledger.

Independent [*stakeholder*](#stakeholder-organization) organizations wishing to transact with one another using DIDs can come together to form a [*consortium*](#consortium) to manage their shared custody of a ledger.

> This spec defines a [discovery service](#discovery-service). The discovery service provided by did:bloc allows a resolver to verify that a consortium is endorsed by its constituent stakeholder organizations, verify that the configuration files of each stakeholder (which includes a list of sidetree endpoints) are signed by the respective stakeholders, and use the provided sidetree endpoints to perform sidetree DID operations.

Reviewer: did:trustbloc

> This spec defines a [discovery service](#discovery-service). The discovery service provided by did:bloc allows a resolver to verify that a consortium is endorsed by its constituent stakeholder organizations, verify that the configuration files of each stakeholder (which includes a list of sidetree endpoints) are signed by the respective stakeholders, and use the provided sidetree endpoints to perform sidetree DID operations.

Reviewer 2: ```suggestion
This spec defines a [discovery service](#discovery-service). The discovery service provided by did:trustbloc allows a client to verify that a consortium is endorsed by its constituent stakeholder organizations, verify that the configuration files of each stakeholder (which includes a list of sidetree endpoints) are signed by the respective stakeholders, and use the provided sidetree endpoints to perform sidetree DID operations.
```

##### Definitions
###### Discovery Service:

> A mechanism that allows resolvers to discover and validate DID operation endpoints.

Reviewer 2: ```suggestion
A mechanism that allows clients to discover and validate DID operation endpoints.
```

###### Consortium:
A group of Stakeholder organizations which all manage DIDs with their own Sidetree nodes, all on the same ledger. A Consortium hosts the configuration files which a resolver uses when first connecting to the network.

> ###### Stakeholder organization:

Reviewer 2: why does it have to be an "organization"? why not a peer? Or just a "stakeholder"?

Reviewer 2: can't Alice running a couple of sidetree nodes be a member of one of these consortiums?

Author: Good point.

##### Consortium Config Files
A consortium config file is a JWS, signed by the stakeholders, with the payload being a JSON object containing:
  - The domain name of the consortium
  - [Consortium policy](#consortium-policy-configuration) configuration settings
  - A list of stakeholders - containing, for each stakeholder:
    - The web domain where its configuration can be found
    - The did:trustbloc DID of the stakeholder, with the associated DID doc in Sidetree on the consortium ledger
    - The verification key which can be used to verify this stakeholder's signature. The key should match the verification key in the stakeholder's DID doc.
  - The hash of the previous version of this config file
  
Example of the format of the configuration data wrapped within the JWS:
```json
{
    "domain": "[consortium domain]",

>     "policy": {},

Reviewer 2: suggest filling out an example policy here

Author: done

    "stakeholders": [
        {
            "domain": "stakeholder.one",
            "did": "[stakeholder one's DID]",

>             "key": "[stakeholder one's verification key]"

Reviewer 2: what's the format of this key?

Author: It needs to be something that's still accessible if the stakeholder's DID doc is revoked, or if the verification key is replaced in the DID doc. I'm thinking did:key.

Author: I've gone with did:key.

        },
        {
            "domain": "stakeholder.two",
            "did": "[stakeholder two's DID]",
            "key": "[stakeholder two's verification key]"
        }
    ],

>     "previous": "[hash of previous consortium config file]"

Reviewer 2: which digest algorithm? sha256?

Reviewer 2: what's the canonicalization scheme? [this one](https://tools.ietf.org/html/draft-rundgren-json-canonicalization-scheme-17)?

Author: No canonicalization necessary, the file is a JWS wrapping a config payload, and can be treated as raw text.

A stakeholder's config file is a JWS, signed by the stakeholder, with the payload being a JSON object containing:
- The stakeholder's domain
- The stakeholder's DID (did:trustbloc)
- Stakeholder custom configuration settings
- The stakeholder's Sidetree endpoints
- the hash of the previous version of this config file

```json
{
    "domain": "stakeholder.one",

>     "did": "did:bloc:[consortium domain]:[stakeholder one's DID]",

Reviewer 2: ```suggestion
    "did": "did:bloc:[consortium domain]:[stakeholder one's DID unique suffix]",
```

>     "did": "did:bloc:[consortium domain]:[stakeholder one's DID]",

Reviewer 2: ```suggestion
    "did": "did:trustbloc:[consortium domain]:[stakeholder one's DID]",
```

Reviewer 2: "stakeholder one's DID" is the last portion of this trustbloc DID. So we have a DID within a DID, eg.:

```
did:trustbloc:myconsortium.com:did:sidetree:1234567abcdefg;initial-values=<encoded-create-operation-request>
```

Can we drop the `did:sidetree:` portion from the embedded DID?

Author: This being an example, I can instead drop the `did:trustbloc:[consortium domain]:` bit, and leave it as saying it's the stakeholder's DID.

Author: Done.

>     "conf": {},

Reviewer 2: what goes in here?

Author: Added a section which describes the contents of `conf`.

`history/` is a directory of previous stakeholder configs for this stakeholder

### Consortium Policy Configuration
The `policy` element of a consortium config object is a JSON object. Each key-value pair is a specific rule for the client to follow when processing consortium or stakeholder configuration files, or when resolving DIDs within the consortium.

##### Caching

> `"cache": {"max-age": [number in seconds]}`

Reviewer 2: what's the size of this number? unsigned 64-bit int?

This element specifies the length of time that a client should cache a copy of the consortium config.

##### Stakeholder Queries

> `""num-queries": [number of queries]`

Reviewer 2: same here - what's the size of this number?

This element specifies the number of stakeholders that a client should query when verifying a consortium configuration - whether for bootstrapping or for updating.

### Bootstrapping Trust

> The Bloc DID method is designed to offer several mechanisms for a resolver to confirm that it can trust the stakeholders to provide the DID doc ledger service. 

Reviewer: TrustBloc DID method

> The Bloc DID method is designed to offer several mechanisms for a resolver to confirm that it can trust the stakeholders to provide the DID doc ledger service. 

Reviewer 2: ```suggestion
The TrustBloc DID method is designed to offer several mechanisms for a client to confirm that it can trust the stakeholders to provide the DID doc ledger service. 
```

In each mechanism, the resolver has the option to cache verified configuration files for a period of time (the cache lifetime being specified in the configuration), and in some of these, the resolver can automatically fetch and verify updates.

> #### Automatic Bootstrapping

Reviewer 2: we should decouple bootstrapping from *did resolution* in this section. Bootstrapping applies to all CRUD operations.

Author: I think my changes have done this now, you can let me know what you think.

- Given a bloc DID: `did:trustbloc:[domain]:[hash]`
- Check local cache for fresh config files from a previous bootstrapping.

> - If there isn't any cached consortium config that matches the DID's domain, the response depends on the configuration of the client - it could abort, or automatically bootstrap trust with the consortium, caching the config.

Reviewer: Should add some explanatory text that after the client aborts, its operator can then decide to examine the validated config and set it trusted. Note: validated means that the client received sufficient matching configurations from different organizations.

- If, after that, there is a cached consortium config for the DID's domain, select a sidetree endpoint from a cached stakeholder and perform the sidetree operation.

> #### CRUD Operations

Reviewer: We should use the term `Method` Operations (rather than CRUD). Also below.
